### PR TITLE
SOLR-14470: Add streaming expressions to /export handler

### PR DESCRIPTION
**Pull request by sigram**
Details in Jira.

Please note that this branch depends on SOLR-14423 (PR 1486) so it appears that there are many unrelated changes.

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java [hunk 1]
```diff
@@ -129,51 +121,46 @@ public StatsStream(StreamExpression expression, StreamFactory factory) throws IO
...
+    expression.addParameter(collection);
 
     // parameters
-    ModifiableSolrParams tmpParams = new ModifiableSolrParams(params);
-
-    for (Entry<String, String[]> param : tmpParams.getMap().entrySet()) {
-      expression.addParameter(new StreamExpressionNamedParameter(param.getKey(),
-          String.join(",", param.getValue())));
+    ModifiableSolrParams mParams = new ModifiableSolrParams(params);
+    for (Entry<String, String[]> param : mParams.getMap().entrySet()) {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java:152]: The ".getMap().entrySet()" portion here is needless; a SolrParams is Iterable.  I'll file a PR to make this change comprehensively as I see it commonly in streaming expressions code.

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java [hunk 2]
```diff
@@ -266,91 +252,106 @@ public void close() throws IOException {
...
 
   public int getCost() {
     return 0;
   }
 
-  @Override
-  public StreamComparator getStreamSort() {
-    return null;
+  private void addStat(Tuple tuple, String field, String stat, Object val) {
+    if(stat.equals("mean")) {
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java:344]: Nitpick: I see this a lot in your changes here -- lack of a space following the "if".  Sometimes you remember, sometimes you don't.

File: solr/solrj/src/java/org/apache/solr/common/params/StreamParams.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.common.params;
+
+/**
+ * Stream Parameters and Properties.
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/common/params/StreamParams.java:20]: Lets say "Streaming Expressions" to be more clear :-)

File: solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java [hunk 4]
```diff
@@ -216,14 +376,53 @@ public void write(OutputStream os) throws IOException {
...
+      streamContext.setRequestParams(params);
+      // nocommit enforce this?
+      streamContext.setLocal(true);
+
+      streamContext.workerID = 0;
+      streamContext.numWorkers = 1;
+      streamContext.setSolrClientCache(initialStreamContext.getSolrClientCache());
+      streamContext.setModelCache(initialStreamContext.getModelCache());
+      streamContext.setObjectCache(initialStreamContext.getObjectCache());
+      streamContext.put("core", req.getCore().getName());
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java:403]: It's consistent with other places that create StreamContext.

File: solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java [hunk 5]
```diff
@@ -285,22 +484,48 @@ protected void addDocsToItemWriter(List<LeafReaderContext> leaves, IteratorWrite
...
+        }
+        if (t.EOF) {
+          break;
+        }
+        writer.add((MapWriter) ew -> t.writeMap(ew));
+      }
+      tupleStream.close();
+    } else {
+      int count = 0;
+      while (count < totalHits) {
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/handler/export/ExportWriter.java:515]: It's the same structure as in `writeDocs(...)`. Yeah, I can change it in both places.

File: solr/core/src/java/org/apache/solr/handler/sql/FilterCalciteConnection.java [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+import org.apache.calcite.linq4j.Enumerator;
+import org.apache.calcite.linq4j.Queryable;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.schema.SchemaPlus;
+
+/**
+ * A filter that contains another {@link CalciteConnection} and
+ * allows adding pre- post-method behaviors.
+ */
+class FilterCalciteConnection implements CalciteConnection {
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/core/src/java/org/apache/solr/handler/sql/FilterCalciteConnection.java:51]: This is no longer needed, not sure how it ended up here.

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java (hunk 2 above)
Comment (sigram) [reply solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/StatsStream.java:344]: It's actually inherited ;) the original style has this all over the place.

File: solr/solrj/src/java/org/apache/solr/common/params/StreamParams.java (hunk 3 above)
Comment (sigram) [reply solr/solrj/src/java/org/apache/solr/common/params/StreamParams.java:20]: +1

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/Tuple.java [hunk 7]
```diff
@@ -44,24 +45,35 @@
...
 
-  public Tuple(){
+  public Tuple() {
     // just an empty tuple
   }
   
-  public Tuple(Map fields) {
-    if(fields.containsKey("EOF")) {
-      EOF = true;
+  public Tuple(Map<?, ?> fields) {
```
[the remark this answers is not in the document]
Comment (sigram) [reply solr/solrj/src/java/org/apache/solr/client/solrj/io/Tuple.java:56]: Added.

**Pull request closed without merging**